pypesto/sample/util.py

```diff
@@ -0,0 +1,147 @@
+import numpy as np
+import logging
+from typing import Tuple
+from tqdm import tqdm
+
+from ..result import Result
+from .diagnostics import geweke_test
+
+logger = logging.getLogger(__name__)
+
+
+def calculate_samples_ci(result: Result,
+                         alpha: float = 0.95
+                         ) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    Calculate parameter confidence intervals based on MCMC samples.
+
+    Parameters
+    ----------
+    result:
+        The pyPESTO result object with filled sample result.
+    alpha:
+        Lower tail probability, defaults to 95% interval.
+
+    Returns
+    -------
+    lb, ub:
+        Bounds of the MCMC percentile-based confidence interval.
+    """
+    # Check if burn in index is available
+    if result.sample_result.burn_in is None:
+        geweke_test(result)
+
+    # Get burn in index
+    burn_in = result.sample_result.burn_in
+
+    # Get converged parameter samples as numpy arrays
+    chain = np.asarray(result.sample_result.trace_x[0, burn_in:, :])
+
+    # Get percentile values corresponding to alpha
+    percentiles = 100 * np.array([(1-alpha)/2, 1-(1-alpha)/2])
+
+    # Get samples' upper and lower bounds
+    lb, ub = np.percentile(chain, percentiles, axis=0)
+
+    return lb, ub
+
+
+def evaluate_samples(result: Result,
+                     stepsize: int = 1
+                     ) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    Evaluate/Simulate MCMC samples.
+
+    Parameters
+    ----------
+    result:
+        The pyPESTO result object with filled sample result.
+    stepsize:
+        Only one in `stepsize` values is simulated for the intervals
+        generation. Recommended for long MCMC chains. Defaults to 1.
+
+    Returns
+    -------
+    evaluation_observables:
+        Simulated model states.
+    evaluation_states:
+        Simulated model observables.
+    """
+
+    # Check if burn in index is available
+    if result.sample_result.burn_in is None:
+        geweke_test(result)
+
+    # Get burn in index
+    burn_in = result.sample_result.burn_in
+
+    # Get full parameter samples as numpy arrays
+    arr_param = np.array(result.sample_result.trace_x[0])
+
+    # Get converged parameter samples as numpy arrays
+    chain = arr_param[np.arange(burn_in, len(arr_param), stepsize)]
+
+    # Number of samples to be simulated
+    nsamples = chain.shape[0]
+
+    # Evaluate median of MCMC samples
```

Comment:
this is unfortunately not ideal, as this depends on the AMICI API. We would need to see whether one can write a proper API for returning data.

Reply:
should this be addressed here or in a different PR?

Reply:
I think here we need to discuss separately first how we want to handle data in pypesto.